File: extractors/python/extract.py
import os
import subprocess
import shutil
from pathlib import Path
import sys

class RepoExtractor:
    def __init__(self, repo_url, output_file="output.txt"):
        self.repo_url = repo_url
        self.output_file = output_file
        self.repo_name = repo_url.split('/')[-1].replace('.git', '')
        self.clone_path = Path(os.getcwd()) / self.repo_name
        self.skip_all = False

    def clone_repository(self):
        """Clone or update the repository."""
        print(f"Starting to process repository: {self.repo_url}")
        
        if not self.clone_path.exists():
            print(f"Cloning repository to: {self.clone_path}")
            result = subprocess.run(['git', 'clone', self.repo_url, str(self.clone_path)], 
                                 capture_output=True, text=True)
        else:
            print(f"Repository already exists at {self.clone_path}. Pulling latest changes.")
            result = subprocess.run(['git', '-C', str(self.clone_path), 'pull', 'origin'],
                                 capture_output=True, text=True)
        
        if result.returncode != 0:
            print("Failed to clone/pull repository:")
            print(result.stderr)
            sys.exit(1)

    def should_include(self, item_path, item_type):
        """Ask user whether to include an item."""
        while True and not self.skip_all:
            response = input(f"Include this {item_type}? (yes/no/folder/skip-all): {item_path}\n").lower()
            
            if response in ['yes', 'no', 'folder', 'skip-all']:
                if response == 'skip-all':
                    self.skip_all = True
                    return 'no'
                return response
            
            print("Invalid response. Please enter 'yes', 'no', 'folder', or 'skip-all'.")
        
        return 'no'

    def get_relative_path(self, full_path):
        """Get path relative to repository root."""
        return str(Path(full_path).relative_to(self.clone_path))

    def write_file_content(self, file_path):
        """Write file content to output file with headers."""
        relative_path = self.get_relative_path(file_path)
        
        with open(self.output_file, 'a', encoding='utf-8') as out:
            out.write(f"----------  FILE: {relative_path}  ----------\n")
            
            try:
                with open(file_path, 'r', encoding='utf-8') as f:
                    out.write(f.read())
            except UnicodeDecodeError:
                print(f"Warning: Could not read {relative_path} as text file")
                
            out.write(f"\n----------  END FILE: {relative_path}  ----------\n\n")

    def process_folder(self, path):
        """Process all files in a folder recursively."""
        if self.skip_all:
            return
            
        print(f"Processing all files within folder {path}")
        for root, _, files in os.walk(path):
            for file in files:
                file_path = Path(root) / file
                if file_path != Path(__file__):
                    print(f"Appending file content: {self.get_relative_path(file_path)}")
                    self.write_file_content(file_path)

    def process_items(self, path):
        """Recursively process folders and files."""
        if self.skip_all:
            return
            
        for item in Path(path).iterdir():
            if item.is_dir():
                response = self.should_include(item, "folder")
                if response == 'yes':
                    self.process_items(item)
                elif response == 'folder':
                    self.process_folder(item)
                elif not self.skip_all:
                    print(f"Skipping folder: {item}")
            else:
                if self.should_include(item, "file") == 'yes' and item != Path(__file__):
                    print(f"Appending file content: {self.get_relative_path(item)}")
                    self.write_file_content(item)
                elif not self.skip_all:
                    print(f"Skipping file: {item}")

    def cleanup(self):
        """Remove cloned repository."""
        print("Cleaning up cloned repository")
        shutil.rmtree(self.clone_path)

    def run(self):
        """Main execution flow."""
        try:
            self.clone_repository()
            print("Starting interactive file selection...")
            self.process_items(self.clone_path)
            print(f"Finished. Check '{self.output_file}'")
        finally:
            self.cleanup()

if __name__ == "__main__":
    if len(sys.argv) < 2:
        print("Usage: python script.py <repo_url> [output_file]")
        sys.exit(1)
    
    repo_url = sys.argv[1]
    output_file = sys.argv[2] if len(sys.argv) > 2 else "output.txt"
    
    extractor = RepoExtractor(repo_url, output_file)
    extractor.run()

File: extractors/javascript/extract.js
const fs = require('fs').promises;
const path = require('path');
const { execSync } = require('child_process');
const readline = require('readline');

class RepoExtractor {
    constructor(repoUrl, outputFile = 'output.txt') {
        this.repoUrl = repoUrl;
        this.outputFile = outputFile;
        this.repoName = repoUrl.split('/').pop().replace('.git', '');
        this.clonePath = path.join(process.cwd(), this.repoName);
        this.skipAll = false;
        
        this.rl = readline.createInterface({
            input: process.stdin,
            output: process.stdout
        });
    }

    async cloneRepository() {
        console.log(`Starting to process repository: ${this.repoUrl}`);
        
        try {
            if (!fs.existsSync(this.clonePath)) {
                console.log(`Cloning repository to: ${this.clonePath}`);
                execSync(`git clone ${this.repoUrl} ${this.clonePath}`);
            } else {
                console.log(`Repository already exists at ${this.clonePath}. Pulling latest changes.`);
                execSync(`git -C ${this.clonePath} pull origin`);
            }
        } catch (error) {
            console.error('Failed to clone/pull repository:', error.message);
            process.exit(1);
        }
    }

    async shouldInclude(itemPath, itemType) {
        if (this.skipAll) return 'no';

        while (true) {
            try {
                const response = await new Promise(resolve => {
                    this.rl.question(
                        `Include this ${itemType}? (yes/no/folder/skip-all): ${itemPath}\n`,
                        answer => resolve(answer.toLowerCase())
                    );
                });

                if (['yes', 'no', 'folder', 'skip-all'].includes(response)) {
                    if (response === 'skip-all') {
                        this.skipAll = true;
                        return 'no';
                    }
                    return response;
                }
                
                console.log("Invalid response. Please enter 'yes', 'no', 'folder', or 'skip-all'.");
            } catch (error) {
                console.error('Error reading input:', error);
                return 'no';
            }
        }
    }

    getRelativePath(fullPath) {
        return path.relative(this.clonePath, fullPath);
    }

    async writeFileContent(filePath) {
        const relativePath = this.getRelativePath(filePath);
        
        try {
            const content = await fs.readFile(filePath, 'utf8');
            await fs.appendFile(
                this.outputFile,
                `----------  FILE: ${relativePath}  ----------\n${content}\n` +
                `----------  END FILE: ${relativePath}  ----------\n\n`
            );
        } catch (error) {
            console.warn(`Warning: Could not read ${relativePath} as text file`);
        }
    }

    async processFolder(folderPath) {
        if (this.skipAll) return;

        console.log(`Processing all files within folder ${folderPath}`);
        const processDir = async (dir) => {
            const entries = await fs.readdir(dir, { withFileTypes: true });
            
            for (const entry of entries) {
                const fullPath = path.join(dir, entry.name);
                if (entry.isFile() && fullPath !== __filename) {
                    console.log(`Appending file content: ${this.getRelativePath(fullPath)}`);
                    await this.writeFileContent(fullPath);
                } else if (entry.isDirectory()) {
                    await processDir(fullPath);
                }
            }
        };

        await processDir(folderPath);
    }

    async processItems(itemPath) {
        if (this.skipAll) return;

        const entries = await fs.readdir(itemPath, { withFileTypes: true });
        
        for (const entry of entries) {
            const fullPath = path.join(itemPath, entry.name);
            
            if (entry.isDirectory()) {
                const response = await this.shouldInclude(fullPath, "folder");
                if (response === 'yes') {
                    await this.processItems(fullPath);
                } else if (response === 'folder') {
                    await this.processFolder(fullPath);
                } else if (!this.skipAll) {
                    console.log(`Skipping folder: ${fullPath}`);
                }
            } else {
                if (await this.shouldInclude(fullPath, "file") === 'yes' && fullPath !== __filename) {
                    console.log(`Appending file content: ${this.getRelativePath(fullPath)}`);
                    await this.writeFileContent(fullPath);
                } else if (!this.skipAll) {
                    console.log(`Skipping file: ${fullPath}`);
                }
            }
        }
    }

    async cleanup() {
        console.log("Cleaning up cloned repository");
        await fs.rm(this.clonePath, { recursive: true, force: true });
        this.rl.close();
    }

    async run() {
        try {
            await this.cloneRepository();
            console.log("Starting interactive file selection...");
            await this.processItems(this.clonePath);
            console.log(`Finished. Check '${this.outputFile}'`);
        } catch (error) {
            console.error('Error during execution:', error);
        } finally {
            await this.cleanup();
        }
    }
}

if (require.main === module) {
    if (process.argv.length < 3) {
        console.log("Usage: node script.js <repo_url> [output_file]");
        process.exit(1);
    }

    const repoUrl = process.argv[2];
    const outputFile = process.argv[3] || "output.txt";

    const extractor = new RepoExtractor(repoUrl, outputFile);
    extractor.run();
}